docs/tutorial/writing-tests/test-case.rst

```diff
@@ -0,0 +1,1148 @@
+    [...]
+
+    ----------------------------------------------------------------------
+    Ran 2 tests in 0.002s
+
+    OK
+
+.. note::
+   You can also run ``python3 manage.py test`` without the ``-u``. Every
+   provider comes with a set of builtin tests like ``shellcheck``
+   (for the ``commands:`` sections) and flake8 (for all ``bin/*.py`` files).
+   Not providing ``-u`` will simply run all tests.
+
+Gathering Coverage from Unit Tests
+----------------------------------
+
+In Checkbox we have a coverage requirement for new pull requests.
+This is to ensure that new contributions do not add source paths that are not
+explored in testing and therefore easy to break down the line with any change.
+
+If you want to collect the coverage of your contribution you can run the
+following:
+
+.. code-block:: none
+
+    (checkbox_venv) > python3 -m coverage run manage.py test -u
+    (checkbox_venv) > python3 -m coverage report --include=bin/*
+    Name                       Stmts   Miss  Cover
+    ----------------------------------------------
+    bin/network_available.py      25     10    60%
+    ----------------------------------------------
+    TOTAL                         25     10    60%
+    (checkbox_venv) > python3 -m coverage report --include=bin/* -m
+    Name                       Stmts   Miss  Cover   Missing
+    --------------------------------------------------------
+    bin/network_available.py      25     10    60%   8-18, 29, 49-52, 56
+    --------------------------------------------------------
+    TOTAL                         25     10    60%
+
+    # You can also get an html report with the following
+    # it is very convenient as you can see file per file what lines are covered
+    # in
+    (checkbox_venv) > python3 -m coverage html
+
+As you can see we are way below the coverage target but this is difficult to
+fix, we should add an end to end test of the main function, so that we
+cover it but, most importantly, we leave trace in the test file of an expected
+usage of the script. Add the following to ``tests/test_network_available.py``
+
+.. code:: python
+
+    class TestMain(unittest.TestCase):
+
+        @mock.patch("subprocess.check_output")
+        def test_nominal(self, check_output_mock):
+            check_output_mock.return_value = textwrap.dedent(
+                """
+                PING 1.1.1.1 (1.1.1.1) from 192.168.1.100 wlan0: 56(84) bytes
+                64 bytes from 1.1.1.1: icmp_seq=1 ttl=53 time=39.0 ms
+                64 bytes from 1.1.1.1: icmp_seq=2 ttl=53 time=143 ms
+
+                --- 1.1.1.1 ping statistics ---
+                2 packets transmitted, 2 received, 0% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            network_available.main(["--threshold", "20", "wlan0"])
+            self.assertTrue(check_output_mock.called)
+
```

Comment:
You can add a conclusion sentence for this section like

> We are now reaching *X%* of test coverage, this is good enough for what we want to achieve.
> 
> Having unit tests for Python scripts used in the Checkbox jobs will help you and others maintain the scripts and prevent regressions in the future.

(or something along the way)

Reply:
I think the paragraph introduction above is pretty clear